### PR TITLE
file existence macros

### PR DESCRIPTION
**Pull request by madhur-tandon**
Assumes that contents of `test.yaml` are:

```
exists:
    site_packages:
        - numpy
    lib:
        - libcurl
    include:
        - xtensor
        - xtensor.hpp
        - xtensor/xarray.hpp
    bin:
        - f2py
        - f2py3
        - curl-config
    cmake_find:
        - xtensor
    pkg_config:
        - xtensor
    file:
        - ssl/cacert.pem
        - ssl/misc
    glob:
        - doc/**/Copyright.txt
```

### CODE REVIEW

File: boa/core/test.py [hunk 1]
```diff
@@ -352,6 +352,26 @@ def construct_metadata_for_test(recipedir_or_package, config):
...
+    PREFIX = "/Users/madhur/mambaforge/envs/boa/"
+    import ruamel.yaml as yaml
+    from rich.console import Console
+
+    console = Console()
+
+    f = open(test_yaml_path)
+    test_yaml = yaml.safe_load(f)
+
+    exists = test_yaml["exists"] if "exists" in test_yaml else None
```
Comment (wolfv) [inline boa/core/test.py:365]: you can use `test_yaml.get("exists")` that will return either the value or `None`.

File: boa/core/test.py [hunk 2]
```diff
@@ -352,6 +352,26 @@ def construct_metadata_for_test(recipedir_or_package, config):
     return m, hash_input
 
 
+def test_run_v2(test_yaml_path,):
+    PREFIX = "/Users/madhur/mambaforge/envs/boa/"
```
Comment (wolfv) [inline boa/core/test.py:356]: PREFIX should be an input to the function

File: boa/core/test.py [hunk 3]
```diff
@@ -352,6 +352,26 @@ def construct_metadata_for_test(recipedir_or_package, config):
...
+    import ruamel.yaml as yaml
+    from rich.console import Console
```
Comment (wolfv) [inline boa/core/test.py:358]: imports shoudl go to the top

File: boa/core/test.py (hunk 1 above)
Comment (madhur-tandon) [reply boa/core/test.py]: Done

File: boa/core/test.py (hunk 2 above)
Comment (madhur-tandon) [reply boa/core/test.py]: Done

File: boa/core/test.py (hunk 3 above)
Comment (madhur-tandon) [reply boa/core/test.py:358]: Will do this when we integrate

**Comment by madhur-tandon** [comment]
Screenshot 2021-09-10 at 2 49 43 PM

**Comment by madhur-tandon** [comment]
Things to consider:

-- setting return code as non-zero when test fails
-- more platform specific checks

File: boa/core/test.py [hunk 4]
```diff
@@ -352,6 +358,217 @@ def construct_metadata_for_test(recipedir_or_package, config):
...
+
+                test_lib = test_lib and test_bin_path and test_path_win
+    return test_lib
+
+
+def check_include(include_dir, include):
+    test_include = True
+    if include:
+        console.print("[blue]- Checking for include[/blue]")
+        for each_include in include:
```
Comment (wolfv) [inline boa/core/test.py:424]: can we do something like:

```py
include_files = [os.path.join(include_dir, fname) for fname in include]
return check_file_existence(include_files)
```

?

Comment (madhur-tandon) [reply boa/core/test.py]: done

**Comment by wolfv** [comment]
for libraries, we should also check the name with prefix (e.g. if you put:

```
library:
- curl
```

it should check for `libcurl`, too.

**Comment by madhur-tandon** [comment]
@wolfv According to the snippet posted in the first comment of this PR, I assumed that one would enter `libcurl` in the `lib` key only. Are we expecting just `curl` instead?

**Pull request merged**